### ISSUE TITLE
macOS app not opening

### ISSUE THREAD

**Issue by DestroyerXyz** [kotatogram/kotatogram-desktop#278]
### Steps to reproduce

1. Copy .app file from dmg to Applications folder.
2. Right click and click open.

### Expected behaviour

The app launches.

### Actual behaviour

macOS says that it cant open the app.
<img width="254" alt="image" src="https://user-images.githubusercontent.com/70191184/147341805-40972496-0c42-4c83-87b7-b3de0529115d.png">

### Operating system

macOS 12.2 Beta (21D5025f)

### Version of Kotatogram Desktop

1.4.6 beta (TD 3.3)

### Installation source

Static binary from official website

**Comment by ilya-fedin** [comment]
Well, we don't have mac hardware, so any help from your side on why this may happen will be highly appreciated

**Comment by DestroyerXyz** [comment]
Well that was an extremely fast reply

**Comment by DestroyerXyz** [comment]
Can you tell me some ways to find the problem?

**Comment by DestroyerXyz** [comment]
You were having some issues with macOS builds, maybe that is related?

**Comment by ilya-fedin** [comment]
I can suggest the Linux way to get helpful logs: try to run the application from terminal. Both the .app and actual binary inside .app.

**Comment by ilya-fedin** [comment]
> You were having some issues with macOS builds, maybe that is related?

Well, 1.4.6 is built with other build method. Previously we used packages from brew, now we're using tdesktop's build scripts. That way offers support for macOS 10.12+ and native ARM support.

**Comment by ilya-fedin** [comment]
The downside is it builds 6 hours on github actions :rofl:

**Comment by DestroyerXyz** [comment]
```
The application cannot be opened for an unexpected reason, error=Error Domain=RBSRequestErrorDomain Code=5 "Launch failed." UserInfo={NSLocalizedFailureReason=Launch failed., NSUnderlyingError=0x6000036c13b0 {Error Domain=NSPOSIXErrorDomain Code=153 "Unknown error: 153" UserInfo={NSLocalizedDescription=Launchd job spawn failed}}}
```

This error in terminal

**Comment by ilya-fedin** [comment]
It's when try to run the .app I guess, what about the actual binary?

**Comment by DestroyerXyz** [comment]
It doesn't let it run for security reasons

**Comment by ilya-fedin** [comment]
Can you overcome that somehow?

**Comment by DestroyerXyz** [comment]
Let me try

**Comment by DestroyerXyz** [comment]
```
zsh: killed     /Users/shreyans/Downloads/Kotatogram.app/Contents/MacOS/Kotatogram

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

Just this after overriding security policy

**Comment by ilya-fedin** [comment]
:sob: just killed?

**Comment by DestroyerXyz** [comment]
Yeah this is the only logging that happens after launching the binary directly

**Comment by DestroyerXyz** [comment]
Maybe because you switched from x86_64 only to universal binary?

**Comment by ilya-fedin** [comment]
macOS is a murderer :see_no_evil:

**Comment by ilya-fedin** [comment]
> Maybe because you switched from x86_64 only to universal binary?

Maybe, I don't know much about macOS internals

**Comment by DestroyerXyz** [comment]
I think you should atleast stop the update rollout as clicking update button in app breaks it

**Comment by ilya-fedin** [comment]
But a person in kotato's telegram group says it works :thinking:

**Comment by DestroyerXyz** [comment]
Is he on apple silicon?

**Comment by ilya-fedin** [comment]
He is on x64 hackintosh, I don't know version though

**Comment by DestroyerXyz** [comment]
works perfecty if I force it to run through rosetta which means something is wrong with the arm64 binary

**Comment by ilya-fedin** [comment]
I wonder what's wrong with arm64 binary, but looks like macOS is not going to tell that

**Comment by DestroyerXyz** [comment]
I found the issue!
Something is going wrong with app signing so only the x86_64 binary is getting correctly signed.
I locally signed the app with `sudo codesign --force -- deep --sign - /Path/to/App.app`
![Uploading Screenshot 2021-12-25 at 1.58.01 AM.png…]()

**Comment by ilya-fedin** [comment]
kotato has no sign...

**Comment by DestroyerXyz** [comment]
`sudo codesign --verify --verbose ./Path/to/App.app`
```
./Kotatogram.app: code object is not signed at all
In architecture: arm64
```

**Comment by DestroyerXyz** [comment]
What do you mean no sign?

**Comment by DestroyerXyz** [comment]
I think you mean the x86_64 binary is signed but just not verified by apple

**Comment by DestroyerXyz** [comment]
I think you mean the x86_64 binary is signed but just not verified by apple

**Comment by ilya-fedin** [comment]
Only if the build process signs it on its own..

**Comment by ilya-fedin** [comment]
We have such a line in actions, without it people were unable to launch kotato on x86_64 at all. Maybe it doesn't change the arm part for some reason.
https://github.com/kotatogram/kotatogram-desktop-builds/blob/master/.github/workflows/mac.yml#L126

**Comment by DestroyerXyz** [comment]
Yes you are correct both the x86_64 and arm64 aren't signed but I found out that arm64 binaries in macos are now killed on start if not signed: https://github.com/golang/go/issues/42684

**Comment by ilya-fedin** [comment]
Am I understand right that the sign that the linker signs automatically can't be used on another x86_64 machine, but can be used on another arm64 machine? So you can't produce an universal binary that will work on both x86_64 and arm64 without a valid sign?

**Comment by DestroyerXyz** [comment]
I didn't understand what you said
Which compiler are you using?

**Comment by ilya-fedin** [comment]
> I didn't understand what you said

From the linked issue:
> indeed, binaries produced by the machine's clang are codesigned, in what looks like the same exact way as the output of codesign -s -.

And then:
> Looks like binaries aren't required to have a signature linked to a Developer ID if they were signed elsewhere. I was able to apply an ad-hoc signature to a binary on my Intel mac and transfer it to my M1 mac and it ran just fine. Appears like binaries just need a signature, doesn't really seem to matter where it came from.

But this 'ad-hoc sign' doesn't work on x86_64, people are unable to run the app even after removing it from quarantine, so `codesign --remove-signature` is used...

> Which compiler are you using?

cmake says it's AppleClang

**Comment by DestroyerXyz** [comment]
An ad-hoc sign is only valid to the local machine it was signed on.
For example, I just ad-hoc signed the kotatogram app to make it work on my machine but it won't work on someone else's.
```
Executable=/Applications/Kotatogram.app/Contents/MacOS/Kotatogram
Identifier=io.github.kotatogram
Format=app bundle with Mach-O universal (x86_64 arm64)
CodeDirectory v=20400 size=900909 flags=0x2(adhoc) hashes=28147+3 location=embedded
Signature=adhoc
Info.plist entries=28
TeamIdentifier=not set
Sealed Resources version=2 rules=13 files=11
Internal requirements count=0 size=12
```

**Comment by ilya-fedin** [comment]
But they're saying in that issue:
> Looks like binaries aren't required to have a signature linked to a Developer ID if they were signed elsewhere. **I was able to apply an ad-hoc signature to a binary on my Intel mac and transfer it to my M1 mac and it ran just fine.** Appears like binaries just need a signature, doesn't really seem to matter where it came from.

**Comment by DestroyerXyz** [comment]
Maybe macs with the same Apple ID allow it

**Comment by DestroyerXyz** [comment]
https://developer.apple.com/documentation/security/seccodesignatureflags/kseccodesignatureadhoc

**Comment by DestroyerXyz** [comment]
Google Chrome's Signature for comparison:
```
Executable=/Applications/Google Chrome Dev.app/Contents/MacOS/Google Chrome Dev
Identifier=com.google.Chrome.dev
Format=app bundle with Mach-O universal (x86_64 arm64)
CodeDirectory v=20500 size=2465 flags=0x12a00(kill,restrict,library-validation,runtime) hashes=66+7 location=embedded
Signature size=9042
Timestamp=Dec 16, 2021 at 9:55:33 AM
Info.plist entries=39
TeamIdentifier=EQHXZ8M8AV
Runtime Version=12.0.0
Sealed Resources version=2 rules=13 files=60
Internal requirements count=1 size=204
```

**Comment by ilya-fedin** [comment]
So that means we can't have arm64 builds?

**Comment by DestroyerXyz** [comment]
No no you can
This guide looks good: https://localazy.com/blog/how-to-automatically-sign-macos-apps-using-github-actions

**Comment by ilya-fedin** [comment]
It seems the guide is about the paid certificate, we're not going to pay anythin

**Comment by DestroyerXyz** [comment]
No its not

**Comment by ilya-fedin** [comment]
"Another prerequisite is an Apple Developer Program subscription."

**Comment by DestroyerXyz** [comment]
Just a minute

**Comment by DestroyerXyz** [comment]
You just have to create a self signed public/private key pair

**Comment by DestroyerXyz** [comment]
You can try https://github.com/patrickpr/YAOG if you use Windows

**Comment by ilya-fedin** [comment]
I don't use Window

**Comment by DestroyerXyz** [comment]
Oh Linux, then maybe try with wine or you would have to figure it out with the openssl cli

**Comment by DestroyerXyz** [comment]
Does @RadRussianRus have a mac?

**Comment by ilya-fedin** [comment]
no

**Comment by RadRussianRus** [comment]
No, I don't.

**Comment by DestroyerXyz** [comment]
Well, need to sleep
Will check this more tomorrow

**Comment by DestroyerXyz** [comment]
I am having exams for a few days so won't have time for this but I would suggest making a new release with only the x86_64 binary (not universal) so the app can at least be functional till this is resolved.

Thanks

**Comment by jacop5** [comment]
I can help you if you need, I have an M1 Macbook Air

**Comment by ilya-fedin** [comment]
I'm not sure what you can help with, the problem seems to be that ARM macos doesn't run unsigned executables.

**Comment by jacop5** [comment]
You mean unsigned arm64 executables?
I can try to build from source code

**Comment by ilya-fedin** [comment]
> You mean unsigned arm64 executables?

Yeah

**Comment by CodeWithShreyans** [comment]
So, what I found so far after some research:
You have to create a self-signed X509 certificate using OpenSSL and bundle the certificate and private key in a PKCS#12 format using OpenSSL.
You somehow have to add the codesigning policy to it (haven't found a way to do this yet).
Then use https://github.com/Apple-Actions/import-codesign-certs to import the cert into the GitHub actions VM.
Then run ``codesign --force --deep -s "NAME_OF_CERTIFICATE" -v /Path/to/Compiled/App.app``

On a side note, almost all the documentation I found online, even on public forums, simply assumed that the dev owns a mac.
#### So basically Apple makes life hell for macOS (and ios) developers if they don't own a mac.

Thanks!

**Comment by ilya-fedin** [comment]
The only reason why there are mac builds was it's easy to build them since tdesktop supports macos and it's enough to run multiple commands on GH actions. Creating a certificate, store it somewhere and then integrate it with GH actions somehow is too much effort for platform that doesn't even want such apps, IMHO.

**Comment by CodeWithShreyans** [comment]
Well, Rosetta works really well but one day (probably about 3-5 years later) it will be deprecated, so an arm64 binary is not at all necessary right now but I think it will be sooner or later.

**Comment by CodeWithShreyans** [comment]
You can try with a macOS VM https://github.com/kholia/OSX-KVM which will be easier but still a good amount of work.
